Fix Research "Open To LAN" Support

gattsuru opened the pull request:
# Changes proposed in this pull request:
- Send packets to IntegratedServer clients, not just DedicatedServer ones, so long as the IntegratedServer has been opened to other users.

I don't like using port number to check IntegratedServer mode, but as far as I can tell, all the seemingly-related stuff like DedicatedServer.onlineMode values are more related to authentication servers.

Review please

gattsuru committed:
Fix Research Open To LAN Support

Actually send packets to clients of an IntegratedServer which is opened up.

Nightenom commented:
Why do you use such a weird structure at all?
Sending network messages in singleplayer is completely fine

Raycoms commented on src/main/java/com/minecolonies/coremod/datalistener/ResearchListener.java:
```diff
@@ -92,7 +93,7 @@ protected void apply(@NotNull final Map<ResourceLocation, JsonElement> object, @
         // But it is possible to reload data packs live, and while not supported it's something to handle.
 
-        if(server instanceof DedicatedServer)
+        if(server instanceof DedicatedServer || (server instanceof IntegratedServer && server.getServerPort() > 0))
```
Yeah, I think just doing a !world.isRemote check here should be just fine imo

gattsuru replied:
I'm not sure if there's a good world available.  The EventHandler case has event.getPlayer().world.isRemote(), but it's already already checked if the event.getPlayer() is a ServerPlayerEntity and returns false anytime the underlying event is fired (since PlayerLoggedInEvent fires from a server initializing a connection to a player).  For ResearchListener, doing something like server.getWorlds() seems like it's not correct, either.

In practice, I think these events only fire on the server side, so if there's no concerns about extraneous network messaging or performance given that they only happen once or twice a play session, can just simplify the checks.

Nightenom replied:
Do not check for sides unless you have to split calls from client and server because of data/logic, network is completely side safe

Raycoms replied:
Oh wait, this is on the server always, if you have the server object, you must be on the server, so a null check should be enough?

gattsuru replied:
Yeah, from a not-crashing perspective the null check's all you need for ResearchListener.  It's a bit of extra RAM and CPU for single-player instances, since they end up rebuilding the tree twice, but that's probably premature optimization.

I've got a few concerns about order-of-actions logic stuff, but the worst case is seeing autostart+instant research announcements triggering twice on server startup and even that may not be possible, so not worth the ugliness.

marchermans reviewed (changes requested):
Use world.isRemote

gattsuru committed:
Remove unnecessary side checks 

Remove unnecessary side checks for ResearchListener and its corresponding research PlayerLoggedInEvent handler.

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -420,11 +421,7 @@ public static void onPlayerEnterWorld(final PlayerEvent.PlayerLoggedInEvent even
             }
 
-            MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
-            if(server instanceof DedicatedServer)
```
If this here was a problem, I'm quite sure you can do player.world.isRemote

gattsuru replied:
Even on dedicated servers (where all players are remote), that always returns false.  Apparently intentionally, despite what one might expect.

It's not an issue, just philosophically bugged me.

Raycoms replied:
No, that's not true, on dedicated servers, there is a "ServerPlayerEntity" and a "ClientPlayerEntity", so depending on the site it's exactly what you want

Raycoms replied:
If it returns always false in this case, this means this code is only called on the server side (still makes sense to do a !player.world.isRemote check)

gattsuru replied:
This is already wrapped in `if (event.getPlayer() instanceof ServerPlayerEntity)`.  I can add the additional check if you want it for code cleanliness purposes.

Raycoms replied:
Then it's not needed at all

gattsuru replied:
Aye.  So it's gone now.
The point was to try and avoid sending the unnecessary message for situations where the client and server shared JVM, rather than sidedness, but if that's not necessary or premature optimization, there's no reason to bother.

marchermans commented:
I checked you code. You can use world.IsRemote 100% of the time. There is no reason todo these weird checks here.

gattsuru committed:
Merge branch 'version/1.16.3' into version/1.16.3-research-lan-mode

gattsuru committed:
Plantation Unlock Fix

Import cleanup, set LetItGrow to the Plantation research properly.

gattsuru committed:
Merge remote-tracking branch 'gattsuru/version/1.16.3-research-lan-mode' into version/1.16.3-research-lan-mode
